assets/terraform/test/resource_nat_policy_test.go

```diff
@@ -0,0 +1,436 @@
+	}
+}
+
+func (o *expectServerRulesOrder) CheckState(ctx context.Context, req statecheck.CheckStateRequest, resp *statecheck.CheckStateResponse) {
+	service := nat.NewService(sdkClient)
+
+	objects, err := service.List(ctx, o.Location, "get", "", "")
+	if err != nil {
+		resp.Error = fmt.Errorf("failed to query server for rules: %w", err)
+		return
+	}
+
+	type ruleWithState struct {
+		Idx   int
+		State int
+	}
+
+	rulesWithIdx := make(map[string]ruleWithState)
+	for idx, elt := range o.RuleNames {
+		rulesWithIdx[fmt.Sprintf("%s-%s", o.Prefix, elt)] = ruleWithState{
+			Idx:   idx,
+			State: 0,
+		}
+	}
+
+	var prevActualIdx = -1
+	for actualIdx, elt := range objects {
+		if state, ok := rulesWithIdx[elt.Name]; !ok {
+			continue
+		} else {
+			state.State = 1
+			rulesWithIdx[elt.Name] = state
+
+			if state.Idx == 0 {
+				prevActualIdx = actualIdx
+				continue
+			} else if prevActualIdx == -1 {
+				resp.Error = fmt.Errorf("rules missing from the server")
+				return
+			} else if actualIdx-prevActualIdx > 1 {
+				resp.Error = fmt.Errorf("invalid rules order on the server")
+				return
+			}
+			prevActualIdx = actualIdx
+		}
+	}
+
+	var missing []string
+	for name, elt := range rulesWithIdx {
+		if elt.State != 1 {
+			missing = append(missing, name)
+		}
+	}
+
+	if len(missing) > 0 {
+		resp.Error = fmt.Errorf("not all rules are present on the server: %s", strings.Join(missing, ", "))
+		return
+	}
+}
+
+type expectServerRulesCount struct {
+	Prefix   string
+	Location nat.Location
+	Count    int
+}
+
+func ExpectServerRulesCount(prefix string, location nat.Location, count int) *expectServerRulesCount {
+	return &expectServerRulesCount{
+		Prefix:   prefix,
+		Location: location,
+		Count:    count,
+	}
+}
+
+func (o *expectServerRulesCount) CheckState(ctx context.Context, req statecheck.CheckStateRequest, resp *statecheck.CheckStateResponse) {
+	service := nat.NewService(sdkClient)
+
+	objects, err := service.List(ctx, o.Location, "get", "", "")
+	if err != nil {
+		resp.Error = fmt.Errorf("failed to query server for rules: %w", err)
+		return
+	}
+
+	var count int
+	for _, elt := range objects {
+		if strings.HasPrefix(elt.Name, o.Prefix) {
+			count += 1
+		}
+	}
+
+	if count != o.Count {
+		resp.Error = UnexpectedRulesError
+		return
+	}
+}
+
+func TestAccPanosNatPolicy(t *testing.T) {
+	t.Parallel()
+
+	nameSuffix := acctest.RandStringFromCharSet(6, acctest.CharSetAlphaNum)
+	prefix := fmt.Sprintf("test-acc-%s", nameSuffix)
+
+	rulesInitial := []string{"rule-1", "rule-2", "rule-3"}
+	rulesReordered := []string{"rule-2", "rule-1", "rule-3"}
+
+	prefixed := func(name string) string {
+		return fmt.Sprintf("%s-%s", prefix, name)
+	}
+
+	withPrefix := func(rules []string) []config.Variable {
+		var result []config.Variable
+		for _, elt := range rules {
+			result = append(result, config.StringVariable(prefixed(elt)))
+		}
+
+		return result
+	}
+
+	device := devicePanorama
+
+	location := locationByDeviceType(device)
```

Comment:
suggestion: we could benefit from explicit initialisation of the location in the context of the test. The cases don't switch a location back and forth, so the function doesn't really help much.

Reply:
We need some way to run tests against panorama and VM series, which have different locations.